src/test/java/com/google/sps/servlets/UserDataServletTest.java

```diff
@@ -0,0 +1,256 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.gson.Gson;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/** Tests the User Data Servlet */
+@RunWith(JUnit4.class)
+public final class UserDataServletTest {
+  private static final String USER_BIO_PROPERTY = "bio";
+  private static final String USER_EMAIL_PROPERTY = "email";
+  private static final String USER_FOUND_PROPERTY = "user-found";
+  private static final String USER_FRIENDS_LIST_PROPERTY = "friends-list";
+  private static final String USER_ID_PROPERTY = "id";
+  private static final String USER_NAME_PROPERTY = "name";
+
+  // Test User info
+  private static final String TEST_USER_NAME = "Tim";
+  private static final String TEST_USER_ID = "123";
+  private static final String TEST_USER_EMAIL = "tim@gmail.com";
+  private static final String TEST_USER_BIO = "Amazing!";
+  private static final String[] TEST_USER_FRIENDS_LIST = new String[]{"321"};
+  private static final String ALTERNATE_TEST_USER_NAME = "John";
+
+  private final Gson gson = new Gson();
+  // Uses a local datastore stored in memory for tests
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private UserDataServlet servletUnderTest;
+  private StringWriter stringWriter;
```

Comment:
Are writers something you can set up in the set up? Worth noting is that tests do not share state (except maybe the in-memory database).

Reply:
Good point, if tests do not share state, then it wouldn't provide any benefits for creating those private instance variables. I'll declare and instantiate them within the tests.